### ISSUE TITLE
Query Parameter for List of Union

### ISSUE THREAD

**Issue by Zarol** [ExpediaGroup/graphql-kotlin#799]
## **Library Version**
GraphQL Kotlin Spring Server - 3.1.0

## **Describe the bug**
A query parameter of type List of a Union type generates a "valid" schema (see SDL below). However the only valid GraphQL query is the following:
```gql
{
  shapes(colors:[]) {
    id
  }
}
```
and attempting to put any object into `colors: []` fails with "Validation error of type WrongType".

## **To Reproduce**
```kotlin
@SpringBootApplication
class Application

fun main(args: Array<String>) {
    runApplication<Application>(*args)
}

@Component
class ShapesQuery : Query {
    suspend fun shapes(colors: List<Colors>): List<Shape> = listOf()
}

data class Shape(val id: String)

interface Colors

data class Red(val red: String): Colors
data class Blue(val blue: String): Colors
```
- Visit Playground and query for shapes.

## **Expected behavior**
It should probably be invalidated at startup as it would appear Unions on input types are not supported in the GraphQL spec (currently?).

## **Generated SDL**
```sdl
schema {
  query: Query
}

"Directs the executor to include this field or fragment only when the `if` argument is true"
directive @include(
    "Included when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Directs the executor to skip this field or fragment when the `if`'argument is true."
directive @skip(
    "Skipped when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Marks the field or enum value as deprecated"
directive @deprecated(
    "The reason for the deprecation"
    reason: String = "No longer supported"
  ) on FIELD_DEFINITION | ENUM_VALUE

"Exposes a URL that specifies the behaviour of this scalar."
directive @specifiedBy(
    "The URL that specifies the behaviour of this scalar."
    url: String!
  ) on SCALAR

union Colors = Blue | Red

type Blue {
  blue: String!
}

type Query {
  shapes(colors: [Colors!]!): [Shape!]!
}

type Red {
  red: String!
}

type Shape {
  id: String!
}
```

**Comment by smyrick** [comment]
Hello @Zarol , thanks for your issue and support on on library.

Sadly, the GraphQL spec does not support unions or interfaces for input yet

https://www.github.com/graphql/graphql-spec/tree/master/rfcs%2FInputUnion.md

**Comment by smyrick** [comment]
See also #195

**Comment by Zarol** [comment]
Is there any way to invalidate at start up rather than generating a bad schema? I see that there is already an error message for something like `fun doSomething(input: Polymorphic)`. I spent a lot of time trying to figure out if it was GraphQL Kotlin or Playground or GraphQL before I saw that the spec did not support it. I imagine it would probably save others some trouble as well.

**Comment by dariuszkuc** [comment]
I agree schema generation should fail in this scenario.

**Comment by dariuszkuc** [comment]
I believe the problem is here -> https://github.com/ExpediaGroup/graphql-kotlin/blob/master/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/types/generateArgument.kt#L32

```kotlin
if (parameter.isInterface() && parameter.isList().not()) {
  throw InvalidInputFieldTypeException(parameter)
}
```

i.e. we check whether parameter is an interface AND is not a list (as list is an interface), proper check should unwrap the type and then validate it -> something along those lines

```kotlin
val parameterType = if (parameter.isList()) {
  parameter.type.getWrappedType()
} else {
  parameter.type
}
if (parameterType.getKClass().isInterface()) {
  throw InvalidInputFieldTypeException(parameter)
}
```

I'll submit a patch tomorrow.

**Comment by smyrick** [comment]
Thanks for catching this as the real issue @dariuszkuc , I was doing all this from my phone so I was a little to quick to close

**Comment by smyrick** [comment]
This was fixed in [3.4.1](https://github.com/ExpediaGroup/graphql-kotlin/releases/tag/3.4.1)
